File: src/main/java/de/pflugradts/passbird/adapter/clipboard/ClipboardService.java
package de.pflugradts.passbird.adapter.clipboard;

import com.google.common.util.concurrent.UncheckedExecutionException;
import com.google.inject.Inject;
import com.google.inject.Singleton;
import de.pflugradts.passbird.application.ClipboardAdapterPort;
import de.pflugradts.passbird.application.configuration.ReadableConfiguration;
import de.pflugradts.passbird.application.failurehandling.FailureCollector;
import de.pflugradts.passbird.application.util.SystemOperation;
import de.pflugradts.passbird.domain.model.transfer.Output;
import io.vavr.control.Try;
import java.util.Objects;

@Singleton
public class ClipboardService implements ClipboardAdapterPort {

    private static final long MILLI_SECONDS = 1000L;

    @Inject
    private FailureCollector failureCollector;
    @Inject
    private SystemOperation systemOperation;
    @Inject
    private ReadableConfiguration configuration;

    private Thread cleanerThread;

    @Override
    public void post(final Output output) {
        this.terminateCleaner();
        Try.run(() -> systemOperation.copyToClipboard(output.getBytes().asString()))
            .onFailure(failureCollector::collectClipboardFailure);
        this.scheduleCleaner();
    }

    private void terminateCleaner() {
        if (Objects.nonNull(cleanerThread)) {
            cleanerThread.interrupt();
        }
    }

    private void scheduleCleaner() {
        if (isResetEnabledInConfiguration()) {
            cleanerThread = new Thread(() -> sleep().onSuccess(x -> systemOperation.copyToClipboard("")));
            cleanerThread.start();
        }
    }

    private Try<Void> sleep() {
        return Try.run(() -> {
            try {
                Thread.sleep(getDelaySecondsFromConfiguration() * MILLI_SECONDS);
            } catch (InterruptedException ex) {
                // rethrow as unchecked exception since Vavr currently prints checked exceptions to std err
                throw new UncheckedExecutionException(ex);
            }
        });
    }

    private boolean isResetEnabledInConfiguration() {
        return configuration.getAdapter().getClipboard().getReset().isEnabled();
    }

    private int getDelaySecondsFromConfiguration() {
        return configuration.getAdapter().getClipboard().getReset().getDelaySeconds();
    }

}

File: src/main/kotlin/de/pflugradts/passbird/adapter/clipboard/ClipboardService.kt
package de.pflugradts.passbird.adapter.clipboard
import de.pflugradts.kotlinextensions.tryCatching
import de.pflugradts.passbird.application.ClipboardAdapterPort
import de.pflugradts.passbird.application.configuration.ReadableConfiguration
import de.pflugradts.passbird.application.failure.ClipboardFailure
import de.pflugradts.passbird.application.failure.reportFailure
import de.pflugradts.passbird.domain.model.transfer.Output
private const val MILLI_SECONDS = 1000L
class ClipboardService constructor(
    private val clipboardGateway: ClipboardGateway,
    private val configuration: ReadableConfiguration,
    private val resetScheduler: ClipboardResetScheduler = ThreadClipboardResetScheduler(),
) : ClipboardAdapterPort {
    private val cleanerLock = Any()
    private var cleanerGeneration = 0L
    override fun post(output: Output) = tryCatching {
        synchronized(cleanerLock) {
            clipboardGateway.copy(output.shell.asString(), nativeToolingEnabled)
            cleanerGeneration += 1
            cleanerGeneration
        }
    }.onFailure {
        reportFailure(ClipboardFailure(it))
    }.onSuccess(::scheduleCleaner).map {
        Unit
    }
    private fun scheduleCleaner(generation: Long) {
        if (isResetEnabled) {
            resetScheduler.schedule(delaySeconds * MILLI_SECONDS, { clearClipboard(generation) }) { error ->
                reportFailure(ClipboardFailure(error))
            }
        }
    }

    private fun clearClipboard(generation: Long) {
        synchronized(cleanerLock) {
            if (cleanerGeneration == generation) {
                tryCatching { clipboardGateway.copy("", nativeToolingEnabled) }.onFailure { reportFailure(ClipboardFailure(it)) }
            }
        }
    }

    private val nativeToolingEnabled: Boolean get() = configuration.adapter.clipboard.nativeTooling.enabled
    private val isResetEnabled: Boolean get() = configuration.adapter.clipboard.reset.enabled
    private val delaySeconds: Int get() = configuration.adapter.clipboard.reset.delaySeconds
}

fun interface ClipboardResetScheduler {
    fun schedule(delayMillis: Long, task: () -> Unit, onFailure: (Exception) -> Unit)
}

class ThreadClipboardResetScheduler : ClipboardResetScheduler {
    override fun schedule(delayMillis: Long, task: () -> Unit, onFailure: (Exception) -> Unit) {
        Thread {
            try {
                Thread.sleep(delayMillis)
                task()
            } catch (error: InterruptedException) {
                onFailure(error)
            } catch (error: IllegalArgumentException) {
                onFailure(error)
            }
        }.start()
    }
}
